### PR TITLE
add initial flow typings for apollo-client

### PR DESCRIPTION
**Pull request by jbaxleyiii**
I've added flow typings for the most commonly used API (I think I got all the critical bits of index.js) as well as updated the build script to package the typings in the npm bundle. This should mean that flow-typed and custom bindings are no longer needed for apollo-client.

TODO:

- [x] Update CHANGELOG.md with your change

### CODE REVIEW

**Comment by stubailo** [comment]
Is there any way to test that these don't get out of date?

**Comment by jbaxleyiii** [comment]
@stubailo not without writing out tests in `js` instead of `ts`. However, I *think* I can break this file apart and put the files next to their source files under `src`. This may make it easier to keep up to date. I'll work on that tonight

**Comment by stubailo** [comment]
Could we maybe compile the tests, then run the compiled `js` tests with Flow and maybe catch some errors?

**Comment by helfer** [comment]
@jbaxleyiii I wouldn't break it apart. But it would be great if there was a way to check if the types are actually correct. For example, how do you know that there are no syntax errors in the flow types right now?

**Comment by jbaxleyiii** [comment]
@helfer flow has a way of writing tests for the types. I can do that instead and configure the CI to verify it 👍

**Comment by helfer** [comment]
That would be great!

**Comment by helfer** [comment]
Any update on this @jbaxleyiii?

**Comment by jbaxleyiii** [comment]
@helfer I didn't want to finish this until I new I would have some time to support the first couple releases in case there are issues with it.

I'm good to do that now so here are a few smoke tests and CI config!

**Comment by jbaxleyiii** [comment]
After this is released, I can release https://github.com/apollographql/react-apollo/pull/695

Now off to merge PRs / close some issues!

File: index.js.flow [hunk 1]
```diff
@@ -0,0 +1,1124 @@
...
+      idValue: IdValue,
+      typeCondition: string,
+      context: ReadStoreContext,
+    ): boolean,
+  }
+
+  declare export class HeuristicFragmentMatcher
+    extends FragmentMatcherInterface {
+    constructor(): this,
+    ensureReady(): Promise<void>,
```
Comment (helfer) [inline index.js.flow]: Seems like I forgot to remove this in an earlier PR. Once I do remove it, how will we make sure that the flow typings stay updated? Is the only way to guarantee it to test every little thing?

**Pull request merged**